core/src/main/kotlin/org/evomaster/core/problem/external/service/httpws/HarvestActualHttpWsResponseHandler.kt

```diff
@@ -0,0 +1,402 @@
+import org.evomaster.core.Lazy
+import org.evomaster.core.logging.LoggingUtil
+import org.evomaster.core.problem.external.service.ApiExternalServiceAction
+import org.evomaster.core.problem.external.service.httpws.param.HttpWsResponseParam
+import org.evomaster.core.problem.external.service.param.ResponseParam
+import org.evomaster.core.problem.util.ParamUtil
+import org.evomaster.core.problem.util.ParserDtoUtil
+import org.evomaster.core.problem.util.ParserDtoUtil.getJsonNodeFromText
+import org.evomaster.core.problem.util.ParserDtoUtil.parseJsonNodeAsGene
+import org.evomaster.core.problem.util.ParserDtoUtil.setGeneBasedOnString
+import org.evomaster.core.problem.util.ParserDtoUtil.wrapWithOptionalGene
+import org.evomaster.core.remote.TcpUtils
+import org.evomaster.core.remote.service.RemoteController
+import org.evomaster.core.search.gene.Gene
+import org.evomaster.core.search.gene.collection.EnumGene
+import org.evomaster.core.search.gene.optional.OptionalGene
+import org.evomaster.core.search.gene.string.StringGene
+import org.evomaster.core.search.service.Randomness
+import org.glassfish.jersey.client.ClientConfig
+import org.glassfish.jersey.client.ClientProperties
+import org.glassfish.jersey.client.HttpUrlConnectorProvider
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.util.concurrent.ConcurrentLinkedQueue
+import javax.annotation.PostConstruct
+import javax.annotation.PreDestroy
+import javax.ws.rs.ProcessingException
+import javax.ws.rs.client.Client
+import javax.ws.rs.client.ClientBuilder
+import javax.ws.rs.client.Entity
+import javax.ws.rs.client.Invocation
+import javax.ws.rs.core.MediaType
+import javax.ws.rs.core.Response
+import kotlin.math.max
+
+class HarvestActualHttpWsResponseHandler {
+
+    // note rc should never be used in the thread of sending requests to external services
+    @Inject(optional = true)
+    private lateinit var rc: RemoteController
+
+    @Inject
+    private lateinit var config: EMConfig
+
+    @Inject
+    private lateinit var randomness: Randomness
+
+    private lateinit var httpWsClient : Client
+
+
+    /**
+     * TODO
+     * to further improve the efficiency of collecting actual responses,
+     * might have a pool of threads to send requests
+     */
+    private lateinit var threadToHandleRequest: Thread
+
+
+    companion object {
+        private val log: Logger = LoggerFactory.getLogger(HarvestActualHttpWsResponseHandler::class.java)
+        private const val ACTUAL_RESPONSE_GENE_NAME = "ActualResponse"
+
+        init{
+            System.setProperty("sun.net.http.allowRestrictedHeaders", "true");
+        }
+    }
+
+    /**
+     * key is actual request based on [HttpExternalServiceRequest.getDescription]
+     * value is an actual response info
+     */
+    private val actualResponses = mutableMapOf<String, ActualResponseInfo>()
+
+    /**
+     * key is description of request with [HttpExternalServiceRequest.getDescription]
```

Comment:
is `actual request` and `description of request` the same thing? or different? ie, are the 2 keys the same?